src/data_source/storage/sql.rs

```diff
@@ -565,24 +616,24 @@ where
             serde_json::to_value(&leaf.leaf().block_header).map_err(|err| QueryError::Error {
                 message: format!("failed to serialize header: {err}"),
             })?;
-        tx.execute_one_with_retries(
-            "INSERT INTO header (height, hash, data) VALUES ($1, $2, $3)",
-            [
+        tx.upsert(
```

Comment:
Changed all inserts/updates to atomic upserts, I just felt with all the new concurrency it was too error prone to try to reason about when it was possible for the same resource to be inserted twice or not.